pkg/media/samplebuilder/samplebuilder.go

```diff
@@ -59,121 +109,199 @@ func (s *SampleBuilder) releasePacket(i uint16) {
-			for j := firstBuffer; j < i; j++ {
-				s.releasePacket(j)
-			}
+	if s.filled.compare(s.active.tail) == slCompareInside {
+		s.active.tail = s.filled.tail
+	}
+
+	var consume sampleSequenceLocation
 
-			return &media.Sample{Data: data, Duration: time.Duration((float64(samples)/float64(s.sampleRate))*secondToNanoseconds) * time.Nanosecond}, s.lastPopTimestamp
+	for i := s.active.head; s.buffer[i] != nil && i < s.active.tail; i++ {
+		if s.depacketizer.IsDetectedFinalPacketInSequence(s.buffer[i].Marker) {
+			consume.head = s.active.head
+			consume.tail = i + 1
+			break
 		}
+		headTimestamp, hasData := s.fetchTimestamp(s.active)
+		if hasData && s.buffer[i].Timestamp != headTimestamp {
+			consume.head = s.active.head
+			consume.tail = i
+			break
+		}
+	}
 
-		p, err := s.depacketizer.Unmarshal(s.buffer[i].Payload)
-		if err != nil {
-			return nil, 0
+	if consume.empty() {
+		return nil
+	}
+
+	if !purgingBuffers && s.buffer[consume.tail] == nil {
+		// wait for the next packet after this set of packets to arrive
+		// to ensure at least one post sample timestamp is known
+		// (unless we have to release right now)
+		return nil
+	}
+
+	sampleTimestamp, _ := s.fetchTimestamp(s.active)
+	afterTimestamp := sampleTimestamp
+
+	// scan for any packet after the current and use that time stamp as the diff point
+	for i := consume.tail; i < s.active.tail; i++ {
+		if s.buffer[i] != nil {
+			afterTimestamp = s.buffer[i].Timestamp
+			break
+		}
+	}
+
+	// the head set of packets is now fully consumed
+	s.active.head = consume.tail
+
+	// prior to decoding all the packets, check if this packet
+	// would end being disposed anyway
+	if s.partitionHeadChecker != nil {
+		if !s.partitionHeadChecker.IsPartitionHead(s.buffer[consume.head].Payload) {
+			s.droppedPackets += consume.count()
+			s.purgeConsumedBuffers()
+			return nil
 		}
+	}
 
+	// merge all the buffers into a sample
+	data := []byte{}
+	for ; consume.head != consume.tail; consume.head++ {
+		p, err := s.depacketizer.Unmarshal(s.buffer[consume.head].Payload)
+		if err != nil {
+			return nil
+		}
 		data = append(data, p...)
 	}
-	return nil, 0
-}
+	samples := afterTimestamp - sampleTimestamp
 
-// Distance between two seqnums
-func seqnumDistance(x, y uint16) uint16 {
-	diff := int16(x - y)
-	if diff < 0 {
-		return uint16(-diff)
+	sample := &media.Sample{
+		Data:               data,
+		Duration:           time.Duration((float64(samples)/float64(s.sampleRate))*secondToNanoseconds) * time.Nanosecond,
+		PacketTimestamp:    sampleTimestamp,
+		PrevDroppedPackets: s.droppedPackets,
 	}
 
-	return uint16(diff)
+	s.droppedPackets = 0
+
+	s.preparedSamples[s.prepared.tail] = sample
+	s.prepared.tail++
+
+	s.purgeConsumedBuffers()
+
+	return sample
 }
 
 // Pop scans s's buffer for a valid sample.
 // It returns nil if no valid samples have been found.
 func (s *SampleBuilder) Pop() *media.Sample {
-	sample, _ := s.PopWithTimestamp()
-	return sample
+	_ = s.buildSample(false)
```

Comment:
not very intuitive that this mutates state but looks like something inherited. might be worth adding a comment describing the intended side effects

Reply:
Done